### ISSUE TITLE
Complement of piece board class and change of movegenerator class

### ISSUE THREAD

**Issue by cjwssb687** [CSC207-2022F-UofT/course-project-chmess#57]
I found some problems when completing the piece and board classes, that is, the method `board.semiValidMovesForColor` uses `piece.generateMoves`, which in turn uses `MoveGenerator` class, but the subclass of `MoveGenerator`, `KingMovesGen ` uses `board.semiValidMovesForColor`. 
Bugs occur in this case:
##k#
#K##
####

Piece and board as entities should not rely on move and movegenerator.
I think it is better to have semiValidMoves and PostMoveValidator in the same layer.

Complement of piece and board:
        Modify the functions that require `Move `and `MoveGenerator ` to elsewhere.
        `board.previousBoard` and other related methods used to implement undo and load games from notation

Modifications to `MoveGenerator`:
        No longer use `piece.generateMoves`

**Comment by David9Nagy** [comment]
Yes, you're right, semiValidMovesForColor currently breaks clean architecture. It would make more sense to just move the whole method to KingMovesGen. That should resolve the problem. Feel free to do so or move it to where you would think is best. You can create a seperate branch from 'demo2' branch to make your edits, as demo2 currently has our latest implementations.

I don't quite understand what you mean in the second part after 'Complement of piece and board:'. Could you explain it further?

**Comment by cherche** [comment]
I strongly agree that there are some very un–Clean Architecture dependencies; I thought the exact same as you when I encountered `Board.semiValidMovesForColor()`.

I'm not sure about `Piece.generateMoves()`. I imagine that moving this abstraction elsewhere would create a code smell (some sort of suboptimal use of object-oriented programming).

**Comment by cherche** [comment]
Before today, I would've say that there's not enough time to refactor, but the project is in an almost-acceptable state. There are many improvements (see some other other issues) that we remarked earlier. If you (or anyone, for that matter) has the time, consider self-assigning.

**Comment by cjwssb687** [comment]
> Yes, you're right, semiValidMovesForColor currently breaks clean architecture. It would make more sense to just move the whole method to KingMovesGen. That should resolve the problem. Feel free to do so or move it to where you would think is best. You can create a seperate branch from 'demo2' branch to make your edits, as demo2 currently has our latest implementations.
> 
> I don't quite understand what you mean in the second part after 'Complement of piece and board:'. Could you explain it further?


the second part after 'Complement of piece and board:' is some changes I planned to do.

**Comment by cjwssb687** [comment]
> Before today, I would've say that there's not enough time to refactor, but the project is in an almost-acceptable state. There are many improvements (see some other other issues) that we remarked earlier. If you (or anyone, for that matter) has the time, consider self-assigning.

Not only does it break the clean architecture, the current code creates a bug: the two kings may use each other's generatemoves.

It's possible to fix this bug without modifying the structure, but I think it's better to change it all together, I'll try it tomorrow.

**Comment by cjwssb687** [comment]
> I strongly agree that there are some very un–Clean Architecture dependencies; I thought the exact same as you when I encountered `Board.semiValidMovesForColor()`.
> 
> I'm not sure about `Piece.generateMoves()`. I imagine that moving this abstraction elsewhere would create a code smell (some sort of suboptimal use of object-oriented programming).


My idea about generatemoves of piece is that creating a movegenerator for only 5 kinds is better than creating it for all 32 pieces.
